--- mac80211.sh.orig	2013-03-20 12:08:29.242841716 -0400
+++ mac80211.sh	2013-03-20 12:10:23.912836514 -0400
@@ -1,6 +1,8 @@
 #!/bin/sh
 append DRIVERS "mac80211"
 
+. /lib/functions/commotion.sh
+
 mac80211_hostapd_setup_base() {
 	local phy="$1"
 	local ifname="$2"
@@ -340,10 +342,43 @@
 		[ -n "$ifname" ] || {
 			[ $i -gt 0 ] && ifname="wlan${phy#phy}-$i" || ifname="wlan${phy#phy}"
 		}
+
+		# check for a Commotion vif/plug
+		network=$(uci_get wireless $vif network)
+		commotion_profile=$(uci_get network $network profile)
+
 		config_set "$vif" ifname "$ifname"
+		if [ -n "$commotion_profile" ]; then
+			commotion_up $ifname $commotion_profile
+			mode=$(commotion_get_mode $ifname)
+			bssid=$(commotion_get_bssid $ifname)
+			ssid=$(commotion_get_ssid $ifname)
+			channel=$(commotion_get_channel $ifname)
+			wpa=$(commotion_get_wpa $ifname)
+			if [ "$wpa" = "true" ]; then 
+				key=$(commotion_get_wpakey $ifname)
+				config_set "$vif" encryption "psk2"
+				uci_set wireless "$vif" encryption "psk2"
+				config_set "$vif" key "$key"
+				uci_set wireless "$vif" key "$key"
+			fi
+			export channel
+			iw dev "$ifname" set channel "$channel"
+			config_set "$device" channel "$channel"
+			uci_set wireless "$vif" channel "$channel"
+			config_set "$vif" mode "$mode"
+			uci_set wireless "$vif" mode "$mode"
+			config_set "$vif" bssid "$bssid"
+			uci_set wireless "$vif" bssid "$bssid"
+			config_set "$vif" ssid "$ssid"
+			uci_set wireless "$vif" ssid "$ssid"
+		else
 
-		config_get mode "$vif" mode
-		config_get ssid "$vif" ssid
+			config_get mode "$vif" mode
+			config_get ssid "$vif" ssid
+		fi
 
 		# It is far easier to delete and create the desired interface
 		case "$mode" in
@@ -442,12 +474,27 @@
 
 		case "$mode" in
 			adhoc)
-				config_get bssid "$vif" bssid
-				config_get ssid "$vif" ssid
+	
+				# check for a Commotion vif/plug
+				network=$(uci_get wireless $vif network)
+				commotion_profile=$(uci_get network $network profile)
+				if [ -n "$commotion_profile" ]; then
+					bssid=$(commotion_get_bssid $ifname)
+					ssid=$(commotion_get_ssid $ifname)
+					wpa=$(commotion_get_wpa $ifname)
+					if [ "$wpa" = "true" ]; then 
+						key=$(commotion_get_wpakey $ifname)
+						encryption="psk2"
+					fi
+				else
+					config_get ssid "$vif" ssid
+					config_get key "$vif" key 1
+					config_get bssid "$vif" bssid
+					config_get encryption "$vif" encryption
+				fi
+
 				config_get beacon_int "$device" beacon_int
 				config_get basic_rate_list "$device" basic_rate
-				config_get encryption "$vif" encryption
-				config_get key "$vif" key 1
 				config_get mcast_rate "$vif" mcast_rate
 				config_get htmode "$device" htmode
 				case "$htmode" in
